utils/arweave.ts

```diff
@@ -0,0 +1,205 @@
+
+  setLikeCoinTxInfo ({
+    fileSize,
+    ipfsHash,
+    txHash
+  }: {
+    fileSize: number
+    ipfsHash: string
+    txHash: string
+  }) {
+    this.fileSize = fileSize
+    this.ipfsHash = ipfsHash
+    this.txHash = txHash
+  }
+
+  setPublicKey (newPubKey: string) {
+    this.pubKey = Buffer.from(newPubKey, 'base64')
+  }
+
+  getPublicKey () {
+    return this.pubKey
+  }
+
+  setAuthToken (token: string) {
+    this.token = token
+  }
+
+  getSigner = () => ({
+    getAddress: () => this.pubKey?.toString(),
+    _signTypedData: async (
+      _domain: never,
+      _types: never,
+      message: { address: string; 'Transaction hash': Uint8Array }
+    ) => {
+      const convertedMsg = Buffer.from(message['Transaction hash']).toString(
+        'base64'
+      )
+      const apiEndpoints = getApiEndpoints()
+      const res = await useFetch(apiEndpoints.API_POST_ARWEAVE_V2_SIGN, {
+        method: 'POST',
+        body: {
+          signatureData: convertedMsg,
+          fileSize: this.fileSize,
+          ipfsHash: this.ipfsHash,
+          txHash: this.txHash
+        },
+        headers: { Authorization: this.token ? `Bearer ${this.token}` : '' }
+      })
+      const { signature } = res.data.value
+      const bSig = Buffer.from(signature, 'base64')
+      // pad & convert so it's in the format the signer expects to have to convert from.
+      const pad = Buffer.concat([
+        Buffer.from([0]),
+        Buffer.from(bSig)
+      ]).toString(
+        'hex'
+      )
+      return pad
+    }
+  })
+
+  _ready () {}
+}
+
+let WebBundlr: any = null
+
+async function getProvider ({
+  fileSize,
+  ipfsHash,
+  txHash,
+  token
+}: {
+  fileSize: number
+  ipfsHash: string
+  txHash: string
+  token: string
+}) {
+  const apiEndpoints = getApiEndpoints()
+  const { data } = await useFetch(apiEndpoints.API_GET_ARWEAVE_V2_PUBLIC_KEY)
+  const { publicKey } = data.value as { publicKey: string }
+  const provider = new Provider({ publicKey, fileSize, ipfsHash, txHash, token })
+  return provider
+}
+
+async function getBundler ({
+  fileSize,
+  ipfsHash,
+  txHash,
+  token
+}: {
+  fileSize: number
+  ipfsHash: string
+  txHash: string
+  token: string
+}) {
+  if (!WebBundlr) {
+    WebBundlr = (await import('@bundlr-network/client')).default
+  }
+  const p = await getProvider({ fileSize, ipfsHash, txHash, token })
+  const { IRYS_URL } = useRuntimeConfig().public
+  const bundlr = new WebBundlr(
+    IRYS_URL,
+    'matic',
+    p
+  )
+  await bundlr.ready()
+  return bundlr
+}
+
+export async function estimateBundlrFilePrice ({
+  fileSize,
+  ipfsHash
+}: {
+  fileSize: number
+  ipfsHash: string
+}) {
+  const apiEndpoints = getApiEndpoints()
+  const { data } = await useFetch(apiEndpoints.API_POST_ARWEAVE_V2_ESTIMATE, {
+    method: 'POST',
+    body: {
+      fileSize,
+      ipfsHash
+    },
+    key: `estimate-${ipfsHash}-${Date.now()}`,
+    headers: {
+      'Cache-Control': 'no-cache',
+      Pragma: 'no-cache'
+    }
```

Comment:
Why do we need these header? File of same size should receive similar estimation result

Reply:
[🐛 Trigger estimateArweaveFee only on upload instead of watch](https://github.com/likecoin/nft-book-press/pull/290/commits/77c84a227fac2849aeb0fbc10f804327a5ddef17)

Reply:
Since @Watch triggers API race condition, it causes cache inconsistency(maybe, not sure), leading to incorrect priceResult. 
To prevent this, we now trigger estimateArweaveFee only on upload instead of watching fileRecords.